### PR TITLE
 [#2039] Support default value semantics - first PR: API changes

### PR DESCRIPTION
**Pull request by shenodaguirguis**
Adding defaultValue and its APIs to Iceberg Types.NestedField to support default values semantics ([issue #2039](https://github.com/apache/iceberg/issues/2039)). 
This is PR 1 of 5. 
For detailed problem and solution check [defaultValueSupport.md](https://github.com/shenodaguirguis/iceberg/blob/master/defaultValueSupport.md)

### CODE REVIEW

File: api/src/main/java/org/apache/iceberg/types/Types.java
Comment (wmoustafa) [inline api/src/main/java/org/apache/iceberg/types/Types.java:505]: I think dealing with null-default as no-default will work, but from an Avro schema point of view, there is a distinction. Do we want to maintain that distinction, or say that default values in Iceberg are non-null default values only? I am okay with the latter as long as we do not end up doing complex checks to disambiguate between null-default and no-default. This will happen if we do not need to distinguish anywhere to begin with.

Comment (wmoustafa) [reply api/src/main/java/org/apache/iceberg/types/Types.java:505]: Also, I am okay with proceeding with this for now and re-evaluating later, but it is something to think about.

Comment (shenodaguirguis) [reply api/src/main/java/org/apache/iceberg/types/Types.java:505]: Thanks for the review @wmoustafa! 
null-default values today are mapped to optional NestedField (with no default). This shall remain the same, given that for optional NestedFields with no defaults, if the field is not manifested then null will be used. In other words, the mapping will be as follows (which shall become clearer in next PR, which handles the schema mapping/conversion):
```
avro field                 ->  iceberg nested field    | what happens when field is not manifested 
------------------------------------------------------------------------------------------------
optional, no default       ->  optional, no default.   | null is used
required, no default       ->  required, no default.   | error
optional, non-null default ->  optional, with default  | default value is used
required, non-null default ->  required, with default  | default value is used
optional, null default     ->  optional, no default    | null is used
required, null default     ->  optional, no default    | null (same as avro's default value) value is used
```

Comment (wmoustafa) [reply api/src/main/java/org/apache/iceberg/types/Types.java:505]: I had the same table in my scratch notes to reason about this :) Yes, agreed it should work as I mentioned, just discussing this from a semantic point of view and code complexity down the line, but seems the latter will not be an issue, and the former should not be an issue as long as the latter is not.

File: api/src/main/java/org/apache/iceberg/types/Types.java
Comment (shardulm94) [inline api/src/main/java/org/apache/iceberg/types/Types.java:462]: Might be better to just move this check into the constructor of `NestedField`.

File: api/src/main/java/org/apache/iceberg/types/Types.java
Comment (shardulm94) [inline api/src/main/java/org/apache/iceberg/types/Types.java:471]: What concrete Java types are supported for this `Object`. Can we add some validation to ensure the default value objects are compatible with their data types?

Comment (shardulm94) [inline api/src/main/java/org/apache/iceberg/types/Types.java:471]: Are default values supported for container types like struct/map/list?

File: api/src/main/java/org/apache/iceberg/types/Types.java
Comment (shenodaguirguis) [reply api/src/main/java/org/apache/iceberg/types/Types.java:462]: makes sense

File: api/src/main/java/org/apache/iceberg/types/Types.java
Comment (shenodaguirguis) [reply api/src/main/java/org/apache/iceberg/types/Types.java:471]: Object Java type should matches that of the Type, added validation to the constructor

File: api/src/main/java/org/apache/iceberg/types/Types.java
Comment (shenodaguirguis) [reply api/src/main/java/org/apache/iceberg/types/Types.java:462]: so, we are using defaultValue = null to represent no-defaultValue. 
The validation was meant to avoid the case when the user calls an API that passes defaultValue and sets the defaultValue to null in error (to avoid behavior that might be unexpected by the user). However, we can say it is legal to set defaultValue to null, which is equivalent to calling and API with no defaultValue, and we leave it to the user's responsibility to user the right APIs and parameters values (with adding java docs regarding defaultValue to the corresponding APIs). Otherwise, the validation will have to take place in the relevant APIs, not the constructor (because at constructor level it is legal) OR, we will have to use a different mechanism to represent no default value. For example, adding a _hasDefault flag, which is not ideal (extra effort to API callers) or use Optional which is not recommended for parameters.

File: api/src/test/java/org/apache/iceberg/types/TestNestedFiledDefaultValues.java
Comment (shardulm94) [inline api/src/test/java/org/apache/iceberg/types/TestNestedFiledDefaultValues.java:26]: Typo: Filed -> Field

File: api/src/main/java/org/apache/iceberg/types/Types.java
Comment (shenodaguirguis) [reply api/src/main/java/org/apache/iceberg/types/Types.java:471]: Avro 1.8 supports default values for record, map, array, and enum
from https://avro.apache.org/docs/1.8.1/spec.html#schema_record
```
avro type | json type | example
-- | -- | --
...
record | object | {"a": 1}
enum | string | "FOO"
array | array | [1]
map | object | {"a": 1}
```

**Review by wmoustafa — APPROVED** [review]

**Comment by shardulm94** [comment]
I feel we should wait for some time to get early feedback from the community to make sure we agree on a high level. This should help reduce churn in our codebase. Most of my comments in https://github.com/apache/iceberg/pull/2496#pullrequestreview-639438937 are valid for the latest set of changes pushed by @shenodaguirguis and I hope they can be addressed before we merge this internally.

File: api/src/main/java/org/apache/iceberg/types/Types.java
Comment (wmoustafa) [inline api/src/main/java/org/apache/iceberg/types/Types.java:483]: Can we eliminate `XType.validateDefaultValue` and have this method call itself recursively?

File: api/src/main/java/org/apache/iceberg/types/Types.java
Comment (wmoustafa) [inline api/src/main/java/org/apache/iceberg/types/Types.java:892]: I am confused why we need to pass a default value here. Are not the parameters of this method already at the level of the Map keys and values? Passing a Map-level concept seems incompatible with what this method is trying to achieve.

Comment (wmoustafa) [reply api/src/main/java/org/apache/iceberg/types/Types.java:892]: Also, this method (while new) is not called anywhere in this patch.

File: api/src/main/java/org/apache/iceberg/types/Types.java
Comment (wmoustafa) [reply api/src/main/java/org/apache/iceberg/types/Types.java:1022]: Answer to this question depends on https://github.com/linkedin/iceberg/pull/63#discussion_r623705887.

File: api/src/main/java/org/apache/iceberg/types/Types.java
Comment (wmoustafa) [reply api/src/main/java/org/apache/iceberg/types/Types.java:708]: Answer to this question depends on https://github.com/linkedin/iceberg/pull/63#discussion_r623705887.

File: api/src/main/java/org/apache/iceberg/types/Types.java
Comment (wmoustafa) [reply api/src/main/java/org/apache/iceberg/types/Types.java:703]: Answer to this question depends on https://github.com/linkedin/iceberg/pull/63#discussion_r623705887.

File: api/src/main/java/org/apache/iceberg/types/Types.java
Comment (wmoustafa) [reply api/src/main/java/org/apache/iceberg/types/Types.java:980]: Answer to this question depends on https://github.com/linkedin/iceberg/pull/63#discussion_r623705887.

File: api/src/main/java/org/apache/iceberg/types/Types.java
Comment (wmoustafa) [reply api/src/main/java/org/apache/iceberg/types/Types.java:1011]: Answer to this question depends on https://github.com/linkedin/iceberg/pull/63#discussion_r623705887.

File: api/src/main/java/org/apache/iceberg/types/Types.java
Comment (shenodaguirguis) [reply api/src/main/java/org/apache/iceberg/types/Types.java:430]: compiler cannot distinguish Object (defaultVlaue) from String (doc) if both APIs have the same name. However, based on our offline discussion, I verified that it is sufficient to maintain the default value in the NestedField level. So, StructType, ListType and MapType changes are revered

File: api/src/main/java/org/apache/iceberg/types/Types.java
Comment (shenodaguirguis) [reply api/src/main/java/org/apache/iceberg/types/Types.java:465]: here, it is a required field and we want to make sure the defaultValue  is not null, because (based on the table [here](https://github.com/linkedin/iceberg/pull/63#discussion_r612030559) this not allowed. In the constructor itself, however, null defaultValue is valid and it means no-defaultValue

File: api/src/main/java/org/apache/iceberg/types/Types.java
Comment (shenodaguirguis) [reply api/src/main/java/org/apache/iceberg/types/Types.java:708]: I am using `hasDefaultValue()` whenever possible (in non-static methods)

File: api/src/main/java/org/apache/iceberg/types/Types.java
Comment (shenodaguirguis) [reply api/src/main/java/org/apache/iceberg/types/Types.java:980]: yes, done

File: api/src/main/java/org/apache/iceberg/types/Types.java
Comment (shenodaguirguis) [reply api/src/main/java/org/apache/iceberg/types/Types.java:892]: that is true. Removed defaultValue from Map/List/Struct level, it is sufficient to maintain it at the NestedField level.

File: api/src/main/java/org/apache/iceberg/types/Types.java
Comment (shenodaguirguis) [reply api/src/main/java/org/apache/iceberg/types/Types.java:483]: done

File: api/src/main/java/org/apache/iceberg/types/Types.java
Comment (shenodaguirguis) [reply api/src/main/java/org/apache/iceberg/types/Types.java:472]: done

File: api/src/main/java/org/apache/iceberg/types/Types.java
Comment (shenodaguirguis) [reply api/src/main/java/org/apache/iceberg/types/Types.java:831]: reverted

File: api/src/main/java/org/apache/iceberg/types/Types.java
Comment (shenodaguirguis) [reply api/src/main/java/org/apache/iceberg/types/Types.java:952]: reverted

**Review by wmoustafa — APPROVED** [review]
LGTM

**Pull request merged**
